src/app/backend/userlinks/userlinks.go

```diff
@@ -0,0 +1,146 @@
+
+	switch {
+	case resource == api.ResourceKindService:
+		return getServiceLinks(client, namespace, name, host)
+	case resource == api.ResourceKindPod:
+		return getPodLinks(client, namespace, name, host)
+	default:
+		log.Printf("Unknown resource types %T!\n", resource)
+	}
+	return
+}
+
+// getServiceLinks get userlinks for services
+func getServiceLinks(client k8sClient.Interface, namespace, name, host string) ([]UserLink, error) {
+	userLinks := []UserLink{}
+	service, err := client.CoreV1().Services(namespace).Get(name, metaV1.GetOptions{})
+	if err != nil || len(service.Annotations[annotationObj]) == 0 {
+		return userLinks, err
+	}
+	m := map[string]string{}
+	err = json.Unmarshal([]byte(service.Annotations[annotationObj]), &m)
+	if err != nil {
+		return userLinks, err
+	}
+
+	for key, uri := range m {
+		userLink := new(UserLink)
+		userLink.Description = key
+		if strings.Contains(uri, apiserverProxyURL) {
+			userLink.Link = host + "/api/v1/namespaces/" + service.ObjectMeta.Namespace +
+				"/services/" + service.ObjectMeta.Name + "/proxy" + strings.TrimPrefix(uri, apiserverProxyURL)
+			userLink.IsURLValid = true
+			userLink.IsProxyURL = true
+		} else if strings.Contains(uri, svcDNSName) {
+			userLink.Link = name + "." + namespace + ".svc.cluster.local" + strings.TrimPrefix(uri, svcDNSName)
+			userLink.IsURLValid = true
+		} else if _, err := url.ParseRequestURI(uri); err != nil {
+			userLink.Link = "Invalid User Link: " + uri
+		} else {
+			if len(service.Status.LoadBalancer.Ingress) > 0 {
+				ingress := service.Status.LoadBalancer.Ingress[0]
+				ip := ingress.IP
+				if ip == "" {
+					ip = ingress.Hostname
+				}
+				for _, port := range service.Spec.Ports {
+					userLink.Link += "http://" + ip + ":" + strconv.Itoa(int(port.Port)) + uri
+					userLink.IsURLValid = true
+				}
+			} else {
+				userLink.Link = uri
+				userLink.IsURLValid = true
+			}
+		}
+		userLinks = append(userLinks, *userLink)
+	}
+	return userLinks, err
+}
+
+// getPodLinks get userlinks for links
+func getPodLinks(client k8sClient.Interface, namespace, name, host string) ([]UserLink, error) {
+	userLinks := []UserLink{}
+	pod, err := client.CoreV1().Pods(namespace).Get(name, metaV1.GetOptions{})
+
+	if err != nil || len(pod.Annotations[annotationObj]) == 0 {
+		return userLinks, err
+	}
+
+	m := map[string]string{}
+	err = json.Unmarshal([]byte(pod.Annotations[annotationObj]), &m)
+	if err != nil {
+		return userLinks, err
+	}
+
+	for key, uri := range m {
+		userLink := new(UserLink)
+		userLink.Description = key
+		if strings.Contains(uri, apiserverProxyURL) {
+			userLink.Link = host + "/api/v1/namespaces/" + pod.ObjectMeta.Namespace +
+				"/pods/" + pod.ObjectMeta.Name + "/proxy" + strings.TrimPrefix(uri, apiserverProxyURL)
+			userLink.IsURLValid = true
+			userLink.IsProxyURL = true
+		} else if strings.Contains(uri, podDNSName) {
+			//10-192-3-8.default.pod.cluster.localhttp://{{pod.dns_name}}:9090/debug
```

Comment:
is this an intended comment? Seems like a good idea, but add some extra words and spaces